src/szk.rs

```diff
@@ -0,0 +1,444 @@
+        nonce: &[u8; 16],
+    ) -> Result<Seed<SEED_SIZE>, SzkError> {
+        let mut xof = P::init(
+            aggregator_blind.as_ref(),
+            &self.adjust_domsep_tag(DST_JOINT_RAND_PART),
+        );
+        // Encode measurement_share (currently an array of field elements) into bytes
+        // and include it in the XOF input.
+        let mut bytes = vec![0u8; T::Field::ENCODED_SIZE];
+        for field_elem in measurement_share {
+            field_elem.encode(&mut bytes)?;
+            xof.update(&bytes)
+        }
+        xof.update(nonce);
+        Ok(xof.into_seed())
+    }
+
+    fn derive_joint_rand_seed(
+        &self,
+        leader_joint_rand_part: &Seed<SEED_SIZE>,
+        helper_joint_rand_part: &Seed<SEED_SIZE>,
+    ) -> Seed<SEED_SIZE> {
+        let mut xof = P::init(
+            &[0; SEED_SIZE],
+            &self.adjust_domsep_tag(DST_JOINT_RAND_SEED),
+        );
+        xof.update(&leader_joint_rand_part.0);
+        xof.update(&helper_joint_rand_part.0);
+        xof.into_seed()
+    }
+
+    fn derive_joint_rands(
+        &self,
+        leader_joint_rand_part: &Seed<SEED_SIZE>,
+        helper_joint_rand_part: &Seed<SEED_SIZE>,
+    ) -> (Seed<SEED_SIZE>, Vec<T::Field>) {
+        let joint_rand_seed =
+            self.derive_joint_rand_seed(leader_joint_rand_part, helper_joint_rand_part);
+        let joint_rands = P::seed_stream(
+            &joint_rand_seed,
+            &self.adjust_domsep_tag(DST_JOINT_RANDOMNESS),
+            &[0],
+        )
+        .into_field_vec(self.typ.joint_rand_len());
+
+        (joint_rand_seed, joint_rands)
+    }
+
+    /// Again the binder is zero here, since the helper's agg_id and the number of proofs doesn't vary.
+    fn derive_helper_proofs_share(&self, proofs_share_seed: &Seed<SEED_SIZE>) -> Vec<T::Field> {
+        Prng::from_seed_stream(P::seed_stream(
+            proofs_share_seed,
+            &self.adjust_domsep_tag(DST_PROOF_SHARE),
+            &[0],
+        ))
+        .take(self.typ.proof_len())
+        .collect()
+    }
+
+    fn derive_query_rands(&self, verify_key: &[u8; SEED_SIZE], nonce: &[u8; 16]) -> Vec<T::Field> {
+        let mut xof = P::init(verify_key, &self.adjust_domsep_tag(DST_QUERY_RANDOMNESS));
+        xof.update(nonce);
+        xof.into_seed_stream()
+            .into_field_vec(self.typ.query_rand_len())
+    }
+
+    fn prove(
+        &self,
+        leader_input_share: &[T::Field],
+        helper_input_share: &[T::Field],
+        encoded_measurement: &[T::Field],
+        rand_seeds: [Seed<SEED_SIZE>; 2],
+        leader_seed_opt: Option<Seed<SEED_SIZE>>,
+        nonce: &[u8; 16],
+    ) -> Result<[SzkProofShare<SEED_SIZE, T::Field>; 2], SzkError> {
+        let [prove_rand_seed, helper_seed] = rand_seeds;
+        let (blind_and_joint_rand_part_for_leader, helper_joint_rand_part, joint_rand) =
+            if self.typ.joint_rand_len() > 0 {
+                let leader_seed = leader_seed_opt.clone().unwrap();
```

Comment:
Rule of thumb: Try to avoid `unwrap()` if possible, as this causes the code to panic. In this situation it's fairly clear that the code will never actually panic, but it's good practice to make your code more readable and free of edge cases.
```suggestion
            if let Some(leader_seed) = leader_seed_opt {
```